pkg/compactor/executor.go

```diff
@@ -283,8 +351,160 @@ func (e *schedulerExecutor) updateJobStatus(ctx context.Context, key *compactors
+		level.Error(e.logger).Log("msg", "invalid compaction plan, abandoning job", "tenant", spec.Tenant)
+		return compactorschedulerpb.ABANDON, errCompactionJobHasNoBlocks
+	}
+
+	userID := spec.Tenant
+	userLogger := util_log.WithUserID(userID, e.logger)
+
+	reg := prometheus.NewRegistry()
+	defer c.syncerMetrics.gatherThanosSyncerMetrics(reg, userLogger)
+
+	userBucket := bucket.NewUserBucketClient(userID, c.bucketClient, c.cfgProvider)
+
+	// Omit ShardAwareDeduplicateFilter. Duplicates are cleaned up during planning via GarbageCollect().
+	// Keep LabelRemoverFilter to remove deprecated labels if applicable.
+	fetcherFilters := []block.MetadataFilter{
+		NewLabelRemoverFilter(compactionIgnoredLabels),
+	}
+
+	var maxLookback = c.cfgProvider.CompactorMaxLookback(userID)
+	if c.cfgProvider.CompactorBlockUploadEnabled(userID) {
+		maxLookback = 0
+	}
+
+	syncDir := c.metaSyncDirForUser(userID)
+	fetcher, err := block.NewMetaFetcher(userLogger, c.compactorCfg.MetaSyncConcurrency, userBucket, syncDir, reg, fetcherFilters, maxLookback)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to create meta fetcher")
+	}
+
+	syncer, err := newMetaSyncer(userLogger, reg, userBucket, fetcher, nil, c.blocksMarkedForDeletion)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to create syncer")
+	}
+
+	blockIDs := make([]ulid.ULID, 0, len(spec.Job.BlockIds))
+	for _, id := range spec.Job.BlockIds {
+		blockID, err := ulid.Parse(string(id))
+		if err != nil {
+			level.Error(userLogger).Log("msg", "invalid block ID, abandoning job", "tenant", userID, "err", err)
+			return compactorschedulerpb.ABANDON, errors.Wrapf(err, "failed to parse block ID")
+		}
+		blockIDs = append(blockIDs, blockID)
+	}
+
+	if err := syncer.SyncRequestedMetas(ctx, blockIDs); err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to sync metas")
+	}
+
+	jobMetas, err := getJobMetas(syncer, blockIDs)
+	if err != nil {
+		level.Error(userLogger).Log("msg", "blocks not found after sync, abandoning job", "err", err)
+		return compactorschedulerpb.ABANDON, err
+	}
+
+	var splitNumShards uint32
+	if spec.Job.Split {
+		splitNumShards = uint32(c.cfgProvider.CompactorSplitAndMergeShards(userID))
+	}
+
+	job, err := buildCompactionJobFromMetas(userID, key.Id, jobMetas, spec, splitNumShards)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, err
+	}
+
+	compactor, err := c.newBucketCompactor(ctx, userID, userLogger, userBucket, syncer, reg)
+	if err != nil {
+		return compactorschedulerpb.REASSIGN, errors.Wrap(err, "failed to create bucket compactor")
+	}
+
+	// Track that a compaction job has started
+	compactor.metrics.groupCompactionRunsStarted.Inc()
+
+	// Clean up the compaction directory. This approach only works while compactors work on a single job at a
+	// time. With higher concurrency, multiple jobs could conflict and clear one anothers progress.
```

Comment:
We don't have any concurrency here though right? Is this a warning if concurrency was introduced?